include/ppx/knob.h

```diff
@@ -0,0 +1,166 @@
+
+class Knob
+{
+public:
+    Knob() {}
+    Knob(const KnobConfig& config, KnobType type)
+        : displayName(config.displayName), flagName(config.flagName), flagDesc(config.flagDesc), type(type) {}
+    virtual ~Knob() {}
+
+    std::string        GetDisplayName() { return displayName; }
+    std::string        GetFlagName() { return flagName; }
+    std::string        GetFlagDesc() { return flagDesc; }
+    KnobType           GetType() { return type; }
+    void               AddChild(Knob* child) { children.push_back(child); }
+    std::vector<Knob*> GetChildren() { return children; }
+
+    virtual void Draw()  = 0;
+    virtual void Reset() = 0;
+
+    // Used for Bool_Checkbox
+    virtual bool GetBoolValue()
+    {
+        PPX_LOG_ERROR("Flag of type " << type << " does not support GetBoolValue()");
+        return false;
+    };
+    virtual void SetBoolValue(bool newVal, bool updateDefault = false)
+    {
+        PPX_LOG_ERROR("Flag of type " << type << " does not support SetBoolValue()");
+    };
+
+protected:
+    std::string        displayName;
+    std::string        flagName;
+    std::string        flagDesc;
+    KnobType           type;
+    std::vector<Knob*> children;
+};
+
+class KnobBoolCheckbox
+    : public Knob
+{
+public:
+    KnobBoolCheckbox() {}
+    KnobBoolCheckbox(const BoolCheckboxConfig& config)
+        : Knob(config, KnobType::Bool_Checkbox)
+    {
+        defaultValue = config.defaultValue;
+        callback     = config.callback;
+    }
+    virtual ~KnobBoolCheckbox() {}
+
+    void Draw() override;
+    void Reset() override;
+
+    bool GetBoolValue() override;
+    void SetBoolValue(bool newVal, bool updateDefault = false) override;
+
+private:
+    bool                      value;
+    bool                      defaultValue;
+    std::function<void(bool)> callback;
+};
+
+class KnobManager
+{
+public:
+    KnobManager() {}
+    virtual ~KnobManager();
+    KnobManager(const KnobManager&)            = delete;
+    KnobManager& operator=(const KnobManager&) = delete;
+
+    // Utilities
+    bool  IsEmpty() { return knobs.empty(); }
+    void  Reset();
+    Knob* GetKnob(int id, bool silentFail = false);
+    Knob* GetKnob(const std::string& name, bool silentFail = false);
+
+    // Create knobs
+    void CreateBoolCheckbox(int i, BoolCheckboxConfig config);
+
+    // Read/Write knobs
+    bool GetKnobBoolValue(int id);
+    void SetKnobBoolValue(int id, bool newVal, bool updateDefault = false);
+
+    // ImGUI
+    void DrawAllKnobs(bool inExistingWindow = false);
+
+    // Command-line flags
+    std::string GetUsageMsg();
+    bool        ParseOptions(std::unordered_map<std::string, CliOptions::Option>& optionsMap);
+
+private:
+    void InsertKnob(int id, Knob* knobPtr);
+    void ConfigureParent(Knob* knobPtr, int parentId);
+    void DrawKnobs(const std::vector<Knob*>& knobsToDraw);
+
+private:
+    std::map<int, Knob*> knobs;
```

Comment:
So you have map, which is sorted by index.
Then you have a tree, because each knob can have children.
Then you have a drawOrder, which is a sorted list giving the sorting order to draw the knobs?

So when you draw you go through the sorted list drawOrder, and recursively go though children to draw them.
So why do you need an ID here, and a sorted map?

If you go with the other suggestion (drop GetKnob etc), and use Knob* as pointers, your drawOders becomes a vector which represents the roots of the tree.
Then, no need to keep all the knobs in the parent, as you already store the pointers in the children.
So if you delete a subtree, it's gone recursively, and you won't have useless knobs left in the manager.

Reply:
Thanks! Looking at the new code now